### ISSUE TITLE
Placeholder error when combining Keras and GPflow

### ISSUE THREAD

**Issue by Zackory** [GPflow/GPflow#728]
I am currently attempting to combine Keras and GPflow, but I am running into an issue with the RBF kernel placeholders when I train the combined model.

The error is:

`
InvalidArgumentError (see above for traceback): You must feed a value for placeholder tensor 'RBF-ead3c7b9-0/variance/initial_unconstrained_value' with dtype double
`

Below is a minimal working example to replicate this error. I am currently running the latest versions of Keras, Tensorflow, and GPflow.

``` python
import os
import numpy as np
import tensorflow as tf
os.environ["KERAS_BACKEND"] = 'tensorflow'
from keras import backend as K
from keras.models import Model
from keras.layers import Input, Dense
from keras.optimizers import Adam
import gpflow

X = np.random.rand(500, 10)
Y = np.random.choice([0, 1], (500, 1), replace=True)
batch_size = 50

netin = Input(shape=(X.shape[1],))
x = Dense(100, activation='relu')(netin)
netout = Dense(1, activation='softmax')(x)

model = Model(inputs=netin, outputs=netout)

x = K.placeholder(shape=(batch_size, X.shape[1]))
y = K.placeholder(shape=(batch_size, Y.shape[1]))

gp_model = gpflow.models.VGP(x, y, kern=gpflow.kernels.RBF(1), likelihood=gpflow.likelihoods.Bernoulli())
loss_model = K.mean(K.binary_crossentropy(y, model(x))) - gp_model.likelihood_tensor
gp_params = tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, "VGP") + tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, 'RBF')

adam = Adam()
param_updates = adam.get_updates(params=model.trainable_weights + gp_params, loss=loss_model)
train_batch = K.function(inputs=[K.learning_phase(), x, y], outputs=[loss_model], updates=param_updates)
updates=gp_param_updates)

num_batches_train = X.shape[0] // batch_size
train_phase = 1
for epoch in range(1, 101):
    # Shuffle data
    inds = np.random.permutation(X.shape[0])
    trainx = X[inds]
    trainy = Y[inds]
    for t in range(num_batches_train):
        l = train_batch([train_phase, trainx[t*batch_size:(t+1)*batch_size], trainy[t*batch_size:(t+1)*batch_size]])
```

**Comment by markvdw** [comment]
@hughsalimbeni: You have mixed NNs with GPflow, any suggestions?

**Comment by hughsalimbeni** [comment]
I think the problem here is that the scoping of 'VGP' isn't actually happening as gpflow appends a random string to the variables names to avoid collisions between models. 

If you take a look at https://github.com/GPflow/GPflow/blob/hugh/advanced_usage_notebook/doc/source/notebooks/advanced_usage.ipynb there are a few pointers how to get these things to work. I haven't actually done this inside a keras model, though.

**Comment by hughsalimbeni** [comment]
also, a problem I've run into sometimes is the model initialization happening in the wrong session. To fix this use `model.enquire_session()` to get the session, rather than using the `tf.Session()`

**Comment by Zackory** [comment]
Thank you for the feedback. I was able to fix this issue by adding the following code:
```python
# Initialize
sess = tf.Session()
K.set_session(sess)
gp_model.initialize(session=sess)
```
I have updated the code above to account for this fix. This code integrates a GPflow model into the loss function of a Keras neural network model.
